Only 1 TensorImpl allocation in differentiable views.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49149 Saves a copy of vector<Tensor> in view ops returning TensorList.
* **#48896 Only 1 TensorImpl allocation in differentiable views.**

Differential Revision: [D25380895](https://our.internmc.facebook.com/intern/diff/D25380895)

Commit:
Only 1 TensorImpl allocation in differentiable views.

[ghstack-poisoned]

Commit:
Update on "Only 1 TensorImpl allocation in differentiable views."




[ghstack-poisoned]

ailzhang commented:
Using test code snippet from @ljk53 's benchmark script:
```
    void view_s1(int count) {
      torch::Tensor a = torch::ones({1, 2, 3});
      torch::Tensor b;
    
      benchmark("view_s1", count, [&]() {
        b = a.view(-1);
      });
    }
```
Before: 722.29ns
After: 583.46ns
(NonVariableType reference: 303.32ns)

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -453,13 +453,22 @@ inline Variable make_variable_differentiable_view(
+        /*version_counter=*/0,
+        /*allow_tensor_metadata_change=*/true);
+    } else {
+       data_impl = data.getIntrusivePtr();
```
Wondering if it is also true that `data.use_count() == 1` here

ailzhang replied:
Great point! That made me realize we were passing Tensor by value which increased the use_count unnecessarily in `as_view`. Updated the condition to rely on use_count. PTAL!

bhosmer reviewed (approved):
Not sure the significance of Ed's `data.use_count() == 1` question, but assuming that isn't flagging something that needs investigation...

Commit:
Update on "Only 1 TensorImpl allocation in differentiable views."




[ghstack-poisoned]

Commit:
Update on "Only 1 TensorImpl allocation in differentiable views."




[ghstack-poisoned]

Commit:
Update on "Only 1 TensorImpl allocation in differentiable views."


Differential Revision: [D25380895](https://our.internmc.facebook.com/intern/diff/D25380895)

[ghstack-poisoned]

ljk53 commented on torch/csrc/autograd/variable.h:
```diff
@@ -453,13 +453,22 @@ inline Variable make_variable_differentiable_view(
+    // moving leaf node into graph interior. This guarantees only 1 TensorImpl
+    // allocation happens in view ops.
+    if (data.getIntrusivePtr().use_count() == 1 && data.getIntrusivePtr()->unique_version()) {
+       data_impl = data.getIntrusivePtr();
```
nit: indent is off, lol

Commit:
Update on "Only 1 TensorImpl allocation in differentiable views."


Differential Revision: [D25380895](https://our.internmc.facebook.com/intern/diff/D25380895)

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -134,7 +134,7 @@ template<typename... Args> inline variable_list flatten_tensor_args(Args&&... ar
 
 // See NOTE [ Autograd View Variables ] for details.
-inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable,
+inline Tensor as_view(const Tensor & base, Tensor& tensor, bool is_differentiable,
```
I'm actually pretty confused; `tensor` looks completely dead in the body here. What's going on?!

ailzhang replied:
it was std::moved into make_variable_differentiable_view below.

ezyang replied:
OK, somehow, I didn't see the site when I was ctrl-f'ing.

Are you sure this new change is right? If the caller to `as_view` expected to retain an owning reference to tensor, the move below is wrong, and taking it by value here was actually the correct thing to do. Hard to say without looking at the call sites. If you do intend to move out of tensor, `Tensor&&` is a more accurate type signature.

ailzhang replied:
@ezyang Hmmm I think the move is not necessary for `tensor` so I just removed it and make them all pass by reference for `tensor`. Local test_autograd run seem to pass for me. Let me know if this looks better now. Thanks!

ezyang replied:
OK, so since you're not moving out, now you can make this a const reference?

swolchok replied:
> If the caller to as_view expected to retain an owning reference to tensor

I don't think this code ended up retaining the passed tensor before or after this change. Did I miss something?

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -453,13 +453,22 @@ inline Variable make_variable_differentiable_view(
+    // moving leaf node into graph interior. This guarantees only 1 TensorImpl
+    // allocation happens in view ops.
+    if (data.getIntrusivePtr().use_count() == 1 && data.getIntrusivePtr()->unique_version()) {
+      data_impl = data.getIntrusivePtr();
```
This does a refcount bump

ailzhang replied:
@ezyang Do you mean we should avoid this refcount bump by using `unsafeGetTensorImpl` etc?

ezyang replied:
Yes. Don't forget that when you do allocate a new shallow copied tesnor you DO need an owning reference in that case

Commit:
Update on "Only 1 TensorImpl allocation in differentiable views."


Differential Revision: [D25380895](https://our.internmc.facebook.com/intern/diff/D25380895)

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -449,17 +449,28 @@ struct TORCH_API DifferentiableViewMeta : public AutogradMeta {
 inline Variable make_variable_differentiable_view(
     Variable base,
-    at::Tensor data,
+    at::Tensor& data,
```
You're not moving out of data anymore, make this a const ref

ezyang commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -194,28 +194,28 @@ inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable
 
 // See NOTE [ Autograd View Variables ] for details.
-inline std::vector<Tensor> as_view(const Tensor & base, std::vector<Tensor> tensors, bool is_differentiable,
+inline std::vector<Tensor> as_view(const Tensor & base, std::vector<Tensor>& tensors, bool is_differentiable,
```
This looks really suspicious. Previously, a copy of `tensors` would have been made when you called `as_view`, so mutating it in the function is harmless. Now you are taking a mutable reference, which means you are silently mutating the caller's copy of tensors. Is this intended?

The old version of this function was implemented in a pretty suspicious way to (do a value-copy on entry, and then mutate this copy, and then return it), but if RVO worked (I don't know if it does in this case), the net number of copies is still 1, which is OK-ish.

ailzhang replied:
@ezyang  I feel mutating the caller's copy is intended. `tensors` are what we eventually return as results and we do want to set up view_meta on these tensors right? So I don't feel bad mutating these tensors since they're supposed to carry this information. 
In the previous implementation we made a copy of vector, but we actually didn't mutate it at all (other than move), but I think move is also redundant.

ailzhang replied:
@swolchok Not sure if this answers your question, the reason is `make_variable_differentiable_view` might either return caller's copy itself or return a new copy so that's why we do 
` tensor = make_variable_non_differentiable_view(base_var, tensor);` below.

swolchok replied:
no, I meant the `tensors` argument to `as_view`. We both mutate it and copy & return it.

ailzhang replied:
oh I see what you mean now, yea we shouldn't copy... 
I'd just return a reference(instead of void) to keep codegen unchanged. fixing...

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -449,17 +449,28 @@ struct TORCH_API DifferentiableViewMeta : public AutogradMeta {
+    // we have to use shallow_copy_and_detach to create a new TensorImpl to avoid
+    // moving leaf node into graph interior. This guarantees only 1 TensorImpl
+    // allocation happens in view ops.
+    if (data.getIntrusivePtr().use_count() == 1 && data.getIntrusivePtr()->unique_version()) {
```
Btw, you changed this due to my comment, but I'm not convinced this version is better than the old one. In particular, you now have to do an atomic load to get the use count; in the previous implementation where you just compared with self, there were no atomics involved. This won't show up in instruction counts but it if there is contention the atomic will matter.

ailzhang replied:
Yea I agree but I checked the change still gives similar perf gain as the direct comparison of tensorimpls. 
One thing I like about `use_count()` is here we make sure we are the only user before we modify it. While in the old implementation (although currently we don't have other users) we might be changing the Tensor while other "users" don't notice. 
I feel the changes are easier to understand than before.

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -449,17 +449,28 @@ struct TORCH_API DifferentiableViewMeta : public AutogradMeta {
+      auto data_impl = data.unsafeGetTensorImpl();
+      data_impl->set_autograd_meta(std::make_unique<DifferentiableViewMeta>(
+        data_impl, std::move(base), std::move(view_func), creation_meta));
+      return Variable(data);
```
Variable doesn't do anything, you don't need to call `Variable()` here

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -449,17 +449,28 @@ struct TORCH_API DifferentiableViewMeta : public AutogradMeta {
+        data_impl, std::move(base), std::move(view_func), creation_meta));
+      return Variable(data);
+    } else {
+      auto data_impl = data.getIntrusivePtr()->shallow_copy_and_detach(
```
The naming here is pretty skeevy, because the type of `data_impl` in the two branches is different. In the top branch, it's non-owning, but this branch, it is owning, that is pretty confusing.

ezyang commented:
IDK, it's getting hairier and hairier. It seems like most of the code here needs to be checked carefully for unnecessary ref count bumps and copy by value. CC'ing @swolchok in case he has some more wisdom.

swolchok commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -194,28 +194,28 @@ inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable
   for(Tensor &tensor : tensors) {
     if (is_differentiable) {
-      tensor = make_variable_differentiable_view(base_var, std::move(tensor), creation_meta);
+      tensor = make_variable_differentiable_view(base_var, tensor, creation_meta);
```
I think not moving here doesn't make a difference. There are two cases to consider:

1) You reuse `tensor`. In this case, whether you move it or not, you don't seem to touch the reference count.
2) You make a new tensor. If you don't move, you have a decref on the old tensor in operator=. If you do move, you have a decref on the old tensor when it goes out of scope in make_variable_differentiable_view.

Commit:
Update on "Only 1 TensorImpl allocation in differentiable views."


Differential Revision: [D25380895](https://our.internmc.facebook.com/intern/diff/D25380895)

[ghstack-poisoned]

ailzhang commented:
Based on the discussion above looks like we are good with using const ref instead of moving to avoid ref bumps? Updated the diff based on the comments, please take another look @ezyang  @swolchok , thanks!

swolchok commented:
Returning a reference will trick callers into copying most likely (they
probably wrote “auto views = as_view...”). Need to change them.

On Wed, Dec 9, 2020 at 4:28 PM Ailing <notifications@github.com> wrote:

> *@ailzhang* commented on this pull request.
> ------------------------------
>
> In torch/csrc/autograd/VariableTypeUtils.h
> <https://github.com/pytorch/pytorch/pull/48896#discussion_r539751137>:
>
> >    }
>  }
>
>  // See NOTE [ Autograd View Variables ] for details.
> -inline std::vector<Tensor> as_view(const Tensor & base, std::vector<Tensor> tensors, bool is_differentiable,
> +inline std::vector<Tensor> as_view(const Tensor & base, std::vector<Tensor>& tensors, bool is_differentiable,
>
> oh I see what you mean now, yea we shouldn't copy...
> I'd just return a reference(instead of void) to keep codegen unchanged.
> fixing...
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/48896#discussion_r539751137>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAIK6E2FNVURZMDWU5XB7DSUAI3XANCNFSM4UO7DKZA>
> .
>
